chips/lowrisc/src/spi_host.rs

```diff
@@ -0,0 +1,672 @@
+    ],
+    err_en [
+        CMDBUSY OFFSET(0) NUMBITS(1) [],
+        OVERFLOW OFFSET(1) NUMBITS(1) [],
+        UNDERFLOW OFFSET(2) NUMBITS(1) [],
+        CMDINVAL OFFSET(3) NUMBITS(1) [],
+        CSIDINVAL OFFSET(4) NUMBITS(1) [],
+    ],
+    err_status [
+        CMDBUSY OFFSET(0) NUMBITS(1) [],
+        OVERFLOW OFFSET(1) NUMBITS(1) [],
+        UNDERFLOW OFFSET(2) NUMBITS(1) [],
+        CMDINVAL OFFSET(3) NUMBITS(1) [],
+        CSIDINVAL OFFSET(4) NUMBITS(1) [],
+        ACCESSINVAL OFFSET(5) NUMBITS(1) [],
+    ],
+    event_en [
+        RXFULL OFFSET(0) NUMBITS(1) [],
+        TXEMPTY OFFSET(1) NUMBITS(1) [],
+        RXWM OFFSET(2) NUMBITS(1) [],
+        TXWM OFFSET(3) NUMBITS(1) [],
+        READY OFFSET(4) NUMBITS(1) [],
+        IDLE OFFSET(5) NUMBITS(1) [],
+    ],
+];
+
+pub struct SpiHost {
+    registers: StaticRef<SpiHostRegisters>,
+    client: OptionalCell<&'static dyn hil::spi::SpiMasterClient>,
+    busy: Cell<bool>,
+    chip_select: Cell<u32>,
+    cpu_clk: Cell<u32>,
+    tsclk: Cell<u32>,
+    tx_buf: TakeCell<'static, [u8]>,
+    rx_buf: TakeCell<'static, [u8]>,
+    tx_len: Cell<usize>,
+    rx_len: Cell<usize>,
+    tx_offset: Cell<usize>,
+    rx_offset: Cell<usize>,
+}
+// SPI Host Command Direction: Bidirectional
+const SPI_HOST_CMD_BIDIRECTIONAL: u32 = 3;
+// SPI Host Command Speed: Standard SPI
+const SPI_HOST_CMD_STANDARD_SPI: u32 = 0;
+
+impl SpiHost {
+    pub const fn new(base: StaticRef<SpiHostRegisters>, cpu_clk: u32) -> Self {
+        SpiHost {
+            registers: base,
+            client: OptionalCell::empty(),
+            busy: Cell::new(false),
+            chip_select: Cell::new(0),
+            cpu_clk: Cell::new(cpu_clk),
+            tsclk: Cell::new(0),
+            tx_buf: TakeCell::empty(),
+            rx_buf: TakeCell::empty(),
+            tx_len: Cell::new(0),
+            rx_len: Cell::new(0),
+            tx_offset: Cell::new(0),
+            rx_offset: Cell::new(0),
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+        let irq = regs.intr_state.extract();
+        self.disable_interrupts();
+
+        if irq.is_set(intr::ERROR) {
+            if self.rx_buf.is_some() {
+                self.rx_buf.take().map(|rx_buf| {
+                    self.client.map(|client| match self.tx_buf.take() {
+                        None => (),
+                        Some(tx_buf) => client.read_write_done(
+                            tx_buf,
+                            Some(rx_buf),
+                            self.tx_offset.get(),
+                            Err(ErrorCode::FAIL),
+                        ),
+                    });
+                });
+                //Specified to be cleared, after err_status is cleared.
+                self.clear_err_interrupt();
+                //Something went wrong, reset IP and clear buffers
+                self.reset_spi_ip();
+                self.reset_internal_state();
```

Comment:
You should probably do this before the callback. It's possible the callback will try the operation again as it failed, in which case you don't want to clear the data after a new operation has started